src/main/java/duke/TaskList.java

```diff
@@ -0,0 +1,158 @@
+     * in the parameters.
+     *
+     * @param tasks The list of tasks.
+     */
+    public TaskList(List<Task> tasks) {
+        this.tasks = tasks;
+    }
+
+    /**
+     * Retrieves the task according to the index given.
+     *
+     * @param i Index of the task in the list.
+     * @return Returns the task.
+     */
+    public Task get(int i) {
+        return tasks.get(i);
+    }
+
+    /**
+     * Retrieves the size of the task list.
+     *
+     * @return Returns the size.
+     */
+    public int size() {
+        return tasks.size();
+    }
+
+    /**
+     * Retrieves the task list.
+     *
+     * @return Returns the list of tasks.
+     */
+    public List<Task> getList() {
+        return this.tasks;
+    }
+
+    /**
+     * Adds a task to the task list.
+     *
+     * @param task Task to be added.
+     */
+    public void addTask(Task task) {
+        tasks.add(task);
+    }
+
+    /**
+     * Marks a task in the task list as complete.
+     *
+     * @param taskNo Index of the task to be completed.
+     * @return Returns the task.
+     */
+    public Task completeTask(int taskNo) {
+        Task task = tasks.get(taskNo - 1);
+        task.markAsDone();
+        return task;
+    }
+
+    /**
+     * Removes a task in the task list.
+     *
+     * @param taskNo Index of the task to be removed.
+     * @return Returns the removed task.
+     */
+    public Task deleteTask(int taskNo) {
+        Task task = tasks.get(taskNo - 1);
+        tasks.remove(task);
+        return task;
+    }
+
+    /**
+     * Edits a task's description in the task list.
+     *
+     * @param taskNo Index of the task to be edited.
+     * @param description New description of the task.
+     * @return Returns the edited task.
+     */
+    public Task updateTaskDesc(int taskNo, String description) {
+        Task task = tasks.get(taskNo - 1);
+        task.updateDesc(description);
+        return task;
+    }
+
+    /**
+     * Edits a task's date in the task list.
+     *
+     * @param taskNo Index of the task to be edited.
+     * @param date New date of the task.
+     * @return Returns the edited task.
+     * @throws DukeInvalidUpdateException when attempting to update an invalid detail.
+     */
+    public Task updateTaskDate(int taskNo, String date) throws DukeInvalidUpdateException {
+        Task task = tasks.get(taskNo - 1);
+        if (task instanceof Event) {
+            Event event = (Event) task;
+            event.updateDate(date);
+        } else if (task instanceof Deadline) {
+            Deadline deadline = (Deadline) task;
+            deadline.updateDate(date);
+        } else {
+            throw new DukeInvalidUpdateException();
+        }
+        return task;
+    }
+
+
+
+    /**
+     * Finds the tasks related to keyword.
+     *
+     * @param keyword Keyword used to find the related tasks.
+     * @return Returns a list of related tasks.
+     */
+    public List<Task> findTasks(String keyword) {
```

Comment:
perhaps this method name could be 'findRelatedTasks' to make it more meaningful, similar to my earlier comments 😄